lib/zip/entry.rb

```diff
@@ -52,25 +53,28 @@ def check_name(name)
       raise ::Zip::EntryNameError, "Illegal ZipEntry name '#{name}', name must not start with /"
     end
 
-    def initialize(*args)
-      name = args[1] || ''
-      check_name(name)
+    def initialize(zipfile = '', name = '', comment = '',
+                   compression_method = DEFLATED,
+                   compression_level = ::Zip.default_compression)
+      @name = name
+      check_name(@name)
 
       set_default_vars_values
       @fstype = ::Zip::RUNNING_ON_WINDOWS ? ::Zip::FSTYPE_FAT : ::Zip::FSTYPE_UNIX
+      @ftype = name_is_directory? ? :directory : :file
 
-      @zipfile            = args[0] || ''
-      @name               = name
-      @comment            = args[2] || ''
-      @extra              = args[3] || ''
-      @compressed_size    = args[4] || 0
-      @crc                = args[5] || 0
-      @compression_method = args[6] || ::Zip::Entry::DEFLATED
-      @size               = args[7] || 0
-      @time               = args[8] || ::Zip::DOSTime.now
+      @zipfile            = zipfile
+      @comment            = comment
+      @compression_method = (@ftype == :directory ? STORED : compression_method)
```

Comment:
I think the ternary here isn't needed, because `@compression_method` is only used if this is a file (and `@compression_level > 0`, which is missing from this check and the one in the setter but present in the getter). It might be clearer to call this something like `@requested_compression_method` to make it clear it's only used in some cases?

Reply:
Agreed that the ternary isn't needed here.

I'll leave the name as it is though, as it does have a meaning for an `Entry` that has been read from a Zip file. So much of `Entry` is confused by the fact that it is used both before and after writing to disk.